### ISSUE TITLE
Upgrading from Teleport 4.1.4 to 4.2.8 does not regenerate host certificates properly

### ISSUE THREAD

**Issue by webvictim** [gravitational/teleport#3654]
**What happened**: I had Teleport 4.1.4 installed on a node, connected back to an auth/proxy via node tunnelling.

I use `HostKeyAlias` in my local SSH config - this means that I can define the CA key for all hosts in my Teleport cluster easily without needing to reference each individual hostname in `~/.ssh/known_hosts` and make changes when I add new nodes. Teleport's default certificate generation method automatically adds `<nodename>.<proxy server address>` as a principal to every host which makes this possible. This is done to make it compatible with the output format from `tctl auth export --type=host`.

`~/.ssh/config`
```
Host zeus
    User gus
    ProxyCommand sh -c '~/.ssh/teleport/login.sh teleport.domain.im username; ssh -i ~/.tsh/keys/teleport.domain.im/username -p 3023 teleport.domain.im -s proxy:%h:%p'
    Port 3022
    IdentityFile ~/.tsh/keys/teleport.domain.im/username
    HostKeyAlias zeus.teleport.domain.im
```

Relevant line in `~/.ssh/known_hosts`:
```
@cert-authority teleport.domain.im,*.teleport.domain.im ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQDX2iIvsuLDdAqd1Kd0LAKa+JRkF/qRNTCqrIuBqjV4F8i0lTQ+8XTr7dqoT1BrzEo39mn8SfjYWyULEGYPwtKr1VXkGhS7Oo6r/jabDuuKNecYFLycxZdbBXItGCOLcbmWwo+xWcHNzmtPa/SERrvsUepS0xsMQR+sOgCy11RQUgpLuy0LDah0toGEeAcmmyZXe0th/11z248GqvfOfGmExbeaVts1pJsSOvUqzO++S2VZmVs0dYiZTYWhu7QUha+yBPowd9X3zzI/gLuxUHat3heqcIfa9WYVwnY9pLD8J+b/ULciOXTwL4KKhk27SjmVqlnwmfWc2FoZc1wlCz/x type=host
```

This means I can run `ssh zeus`, it expands this to `zeus.teleport.domain.im`, the host key is automatically validated against the principals presented by the host, then the process of logging into Teleport and connecting to the node via the Teleport proxy is totally automated for me.

I updated my node to Teleport 4.2.8 and restarted Teleport. As a result of this, I think the node's certificates were regenerated, as per this log:

```
May 01 11:02:12 zeus teleport[4132841]: INFO [PROC:1]    Service Node has updated principals to ["zeus" "9efd8ca9-52e6-4fa4-a6dd-93bdeda99cf4"], DNS Names to [], going to request new principals and update. service/connect.go:680
```

After this, I can't automatically connect to the node using the old method any more:

```
$ ssh -vv zeus
Certificate invalid: name is not a listed principal
debug1: No matching CA found. Retry with plain key
The authenticity of host 'zeus.teleport.domain.im (<no hostip for proxy command>)' can't be established.
RSA key fingerprint is SHA256:jSSSEn15UArsuZFNPTvLg3xG7732DMsK3sY4onnIL0E.
Are you sure you want to continue connecting (yes/no/[fingerprint])? ^C
```

The SSH debug output seems to confirm that the `zeus.teleport.domain.im` principal is no longer present on the certificate:

```
debug1: Server host certificate: ssh-rsa-cert-v01@openssh.com SHA256:jSSSEn15UArsuZFNPTvLg3xG7732DMsK3sY4onnIL0E, serial 0 ID "" CA ssh-rsa SHA256:1paZOuzvmf376q20ZZtfmedb0tS7yYGAGoetMh572aQ valid after 2020-05-01T11:01:12
debug2: Server host certificate hostname: 9efd8ca9-52e6-4fa4-a6dd-93bdeda99cf4.teleport.domain.im
debug2: Server host certificate hostname: zeus
debug2: Server host certificate hostname: 9efd8ca9-52e6-4fa4-a6dd-93bdeda99cf4
debug1: using hostkeyalias: zeus.teleport.domain.im
debug3: hostkeys_foreach: reading file "/home/gus/.ssh/known_hosts"
debug3: record_hostkey: found ca key type RSA in file /home/gus/.ssh/known_hosts:569
debug3: load_hostkeys: loaded 1 keys from zeus.teleport.domain.im
debug1: Host 'hades.teleport.domain.im' is known and matches the RSA-CERT host certificate.
debug1: Found CA key in /home/gus/.ssh/known_hosts:569
```

If I connect to the host, stop Teleport, delete `/var/lib/teleport` and restart Teleport again (so that the host rejoins the cluster), its certificate is regenerated once more and then the connection process works as expected:

```
debug1: Server host certificate: ssh-rsa-cert-v01@openssh.com SHA256:hpRn8cKquIRk/NR/YjnzRKbWShT9aswztz7fPX7FMhU, serial 0 ID "" CA ssh-rsa SHA256:1paZOuzvmf376q20ZZtfmedb0tS7yYGAGoetMh572aQ valid after 2020-05-01T14:13:23
debug2: Server host certificate hostname: 09a498a6-3da5-4a33-a6de-56d1ee6be4cf.teleport.domain.im
debug2: Server host certificate hostname: zeus.teleport.domain.im
debug2: Server host certificate hostname: zeus
debug2: Server host certificate hostname: 09a498a6-3da5-4a33-a6de-56d1ee6be4cf
debug1: using hostkeyalias: zeus.teleport.domain.im
debug1: Host 'zeus.teleport.domain.im' is known and matches the RSA-CERT host certificate.
debug1: Found CA key in /home/gus/.ssh/known_hosts:569
```

Node config:
```yaml
teleport:
  nodename: zeus
  log:
    output: stderr
    severity: INFO
  auth_token: <redacted>
  ca_pin: sha256:<redacted>
  auth_servers: 
  - proxy:3080
auth_service:
  enabled: no
ssh_service:
  enabled: yes
proxy_service:
  enabled: no
```

Auth/proxy config:
```yaml
teleport:
  nodename: artemis
  data_dir: /var/lib/teleport
  pid_file: /var/run/teleport.pid
  log:
    output: stderr
    severity: INFO
auth_service:
  enabled: yes
  cluster_name: teleport.domain.im
  public_addr: teleport.domain.im:3025
  authentication:
    type: github
  tokens:
  - "node:<redacted>"
ssh_service:
  enabled: yes
  public_addr: artemis.teleport.domain.im:3022
proxy_service:
  enabled: yes
  ssh_public_addr: teleport.domain.im:3023
  public_addr: teleport.domain.im:3080
  tunnel_listen_addr: 0.0.0.0:3080
  tunnel_public_addr: teleport.domain.im:3080
  https_key_file: /etc/letsencrypt/live/teleport.domain.im/privkey.pem
  https_cert_file: /etc/letsencrypt/live/teleport.domain.im/fullchain.pem
```

**What you expected to happen**: If we regenerate a host certificate automatically as part of a node upgrade, it should have exactly the same principals as it would if you were to delete all the certs and force a regeneration when joining the cluster as a fresh node.

**How to reproduce it (as minimally and precisely as possible)**: See above.

### Environment
- Teleport version (use `teleport version`): `Teleport v4.2.8 git:v4.2.8-0-ga9015b33 go1.13.2`
- Tsh version (use `tsh version`): Not relevant
- OS (e.g. from `/etc/os-release`): `Fedora 31`

- Where are you running Teleport? (e.g. AWS, GCP, Dedicated Hardware): Home lab

**Comment by webvictim** [comment]
I think this will basically affect any version of Teleport >=4.2.4 (when UUID-based routing was added)

**Comment by zmb3** [comment]
@webvictim do you know if this is still an issue?

**Comment by webvictim** [comment]
I think @espadolini did a bunch of work on this recently to add `nodename` to generated certs and it might have included `ssh_service.public_addr`. It's _probably_ fixed, but worth checking with him?

**Comment by espadolini** [comment]
Fixed by #12272.

**Comment by espadolini** [comment]
Actually now that I'm rereading the description, it might not even be a relevant bug anymore, as even before #12272 we were either adding both `nodename` and `nodename.clustername` or neither - still, now we are doing that.
